File: .superpowers/inline-neuromorphic/frontend/src/services/activity.ts
import api from "./api";

export interface ActivityEvent {
  id: string;
  user_id: string;
  actor_type: string;
  actor_user_id: string | null;
  source: string;
  event_type: string;
  target_type: string;
  file_id: string | null;
  folder_id: string | null;
  share_id: string | null;
  file_request_id: string | null;
  api_token_id: string | null;
  status: number | null;
  ip_address: string | null;
  user_agent: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface ActivityListParams {
  cursor?: string | null;
  limit?: number;
  source?: string;
  event_type?: string;
  target_type?: string;
  file_id?: string;
  folder_id?: string;
  share_id?: string;
  file_request_id?: string;
  api_token_id?: string;
  date_from?: string;
  date_to?: string;
}

export interface ActivityListResponse {
  events: ActivityEvent[];
  next_cursor: string | null;
}

function compactParams(params: ActivityListParams = {}) {
  return Object.fromEntries(
    Object.entries(params).filter(
      ([, value]) => value !== undefined && value !== null && value !== "",
    ),
  );
}

export const activityService = {
  async list(params: ActivityListParams = {}): Promise<ActivityListResponse> {
    const response = await api.get<ActivityListResponse>("/api/activity", {
      params: compactParams(params),
    });
    return response.data;
  },

  async listFile(
    fileId: string,
    params: Omit<ActivityListParams, "file_id"> = {},
  ): Promise<ActivityListResponse> {
    const response = await api.get<ActivityListResponse>(
      `/api/files/${fileId}/activity`,
      { params: compactParams(params) },
    );
    return response.data;
  },
};

File: .superpowers/inline-neuromorphic/backend/src/types/activity.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use serde_json::Value;
use uuid::Uuid;

use crate::entities::audit_event::AuditEvent;

#[derive(Debug, Serialize)]
pub struct ActivityEventResponse {
    pub id: Uuid,
    pub user_id: Uuid,
    pub actor_type: String,
    pub actor_user_id: Option<Uuid>,
    pub source: String,
    pub event_type: String,
    pub target_type: String,
    pub file_id: Option<Uuid>,
    pub folder_id: Option<Uuid>,
    pub share_id: Option<Uuid>,
    pub file_request_id: Option<Uuid>,
    pub api_token_id: Option<Uuid>,
    pub status: Option<i32>,
    pub ip_address: Option<String>,
    pub user_agent: Option<String>,
    pub metadata: Value,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ActivityListResponse {
    pub events: Vec<ActivityEventResponse>,
    pub next_cursor: Option<String>,
}

impl From<AuditEvent> for ActivityEventResponse {
    fn from(event: AuditEvent) -> Self {
        Self {
            id: event.id,
            user_id: event.user_id,
            actor_type: event.actor_type,
            actor_user_id: event.actor_user_id,
            source: event.source,
            event_type: event.event_type,
            target_type: event.target_type,
            file_id: event.file_id,
            folder_id: event.folder_id,
            share_id: event.share_id,
            file_request_id: event.file_request_id,
            api_token_id: event.api_token_id,
            status: event.status,
            ip_address: event.ip_address,
            user_agent: event.user_agent,
            metadata: event.metadata,
            created_at: event.created_at,
        }
    }
}
